_blogposts/compiler/overview-of-new_encoding.mdx

```diff
@@ -0,0 +1,229 @@
+
+Note since V8 or other JavaScript engines are tweaked to make idiomatic JS code running fast, so it also results in faster running code. 
+
+Another property, for a compiled language like BuckleScript, is that we can reason about [IC](https://en.wikipedia.org/wiki/Inline_caching) friendliness by just looking at the type definitions locally; this is very helpful for advanced users to write performance reliable JS code, if you are unfamiliar with how IC works in general, feel free to skip the section about IC.
+
+Note this article is quite dense so that we will skip the old encoding.
+
+## Record (stable)
+
+We compile records to idiomatic JS objects since [version 7](https://bucklescript.github.io/blog/2019/11/18/whats-new-in-7), this is great for performance and debugging. We also support label renaming to shorten field names to save sizes.
+
+Take the code below for example:
+
+```reasonml
+type int64 = {
+    loBits : int [@bs.as "lo"],
+    hiBits : int [@bs.as "hi]
+}
+let value = {hiBits : 33 , loBits : 32 };
+let rand = ({loBits; hiBits}) => loBits + hiBits;
+```
+
+It will generate JS output as below:
+```js
+var value = {lo : 32, hi : 33}
+function rand (param){
+    return param.lo + param.hi
+}
+```
+
+If users want to make it even  shorter, they can choose to compile record as an array as below:
+
+```reasonml
+type int64 = {
+    loBits : int [@bs.as "0"],
+    hiBits : int [@bs.as "1"]
+}
+let value = {hiBits : 33 , loBits : 32 };
+let rand = ({loBits; hiBits}) => loBits + hiBits;
+```
+
+Now output JS as below:
+
+```js
+var value = [32,33]
+function rand(param){
+    return param[0] + param[1]
+}
+```
+
+The label renaming techniques can be applied systematically using a syntactic macro, in the future we may provide an advanced mode to apply it automatically. Another nice property is that only the type definition needs to be adapted, other parts of code is untouched.
+
+### IC friendliness
+
+Record is always in a perfect IC position, the compiler can ensure all generated records are of the same shape
+
+## Variant (internal)
+This encoding for variants may be  subject to change in the future, but it is so simple that it makes sense for users to have a basic understanding.
+
+Take such type definition for example:
+
+```reasonml
+type t =
+  | Black(t, int, t)
+  | Red(t, int, t)
+  | Empty;
+let empty = Empty  ;
+let v0 = Black (empty, 3, empty);
+let v1 = Red (empty, 3, empty);
+```
+The generated JS code would be:
+
+```js
+var empty =/*Empty*/ 0;
+var v0  = {TAG : 0/*Black*/, _0 : /*Empty*/ 0 , _1 : 3 , _2 : /*Empty */ 0};
+var v1 = {TAG : 1/*Red*/, _0 : /*Empty*/ 0 , _1 : 3 , _2 : /*Empty */ 0};
+```
+
+As you can see, variants are divided into two categories, the variant which does not have payload is compiled into a number starting from 0, while variants which has the payload is compiled into an object which has the first slot named `TAG` and the following slots named as `_0`, `_1` ..
```

Comment:
...divided into two categories. Variants which do not have a payload are compiled into a number starting from 0, while variants which have a payload are compiled into an object...